// ng-block-ui/dist/lib/directives/block-ui.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var block_ui_content_component_1 = require("../components/block-ui-content/block-ui-content.component");
var block_ui_default_name_constant_1 = require("../constants/block-ui-default-name.constant");
var BlockUIDirective = (function () {
    function BlockUIDirective(viewRef, templateRef, componentFactoryResolver) {
        this.viewRef = viewRef;
        this.templateRef = templateRef;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    Object.defineProperty(BlockUIDirective.prototype, "blockUI", {
        set: function (name) { this.blockTarget = name; },
        enumerable: true,
        configurable: true
    });
    ;
    BlockUIDirective.prototype.ngOnInit = function () {
        try {
            this.viewRef.createEmbeddedView(this.templateRef);
            var parentElement = this.viewRef.element.nativeElement.nextSibling;
            if (parentElement && !this.isComponentInTemplate(parentElement)) {
                parentElement.classList.add('block-ui__element');
                this.blockUIComponentRef = this.createComponent();
                var blockUIContent = this.findContentNode(this.viewRef.element.nativeElement);
                if (blockUIContent) {
                    parentElement.appendChild(blockUIContent);
                    this.blockUIComponentRef.instance.name = this.blockTarget || block_ui_default_name_constant_1.BlockUIDefaultName;
                    this.blockUIComponentRef.instance.className = 'block-ui-wrapper--element';
                }
            }
        }
        catch (error) {
            console.error('ng-block-ui:', error);
        }
    };
    BlockUIDirective.prototype.isComponentInTemplate = function (element) {
        var children = (element || []).children;
        children = Array.from(children).reverse();
        return children.some(function (el) { return el.localName === 'block-ui'; });
    };
    // Needed for IE (#17)
    BlockUIDirective.prototype.findContentNode = function (element) {
        var nextSibling = element.nextSibling;
        return [nextSibling, nextSibling.nextSibling].find(function (e) { return e.localName === 'block-ui-content'; });
    };
    BlockUIDirective.prototype.createComponent = function () {
        var resolvedBlockUIComponent = this.componentFactoryResolver.resolveComponentFactory(block_ui_content_component_1.BlockUIContentComponent);
        return this.viewRef.createComponent(resolvedBlockUIComponent);
    };
    return BlockUIDirective;
}());
BlockUIDirective.decorators = [
    { type: core_1.Directive, args: [{ selector: '[blockUI]' },] },
];
/** @nocollapse */
BlockUIDirective.ctorParameters = function () { return [
    { type: core_1.ViewContainerRef, },
    { type: core_1.TemplateRef, },
    { type: core_1.ComponentFactoryResolver, },
]; };
BlockUIDirective.propDecorators = {
    'blockUI': [{ type: core_1.Input },],
};
exports.BlockUIDirective = BlockUIDirective;
